### PR TITLE
#4715 edit vaccination transact data via vaccination event modal

### LINKED ISSUE #4715: Add the ability to edit vaccination transact data via the Vaccination Event modal
## Is your feature request related to a problem? Please describe.

Trello: https://trello.com/c/J4kCLZGC/74-17b-17c-edit-vaccine-type

Editing this portion of the 'vaccination event record':
image

## Describe the solution you'd like

It's possible to just edit the name note itself - but we wouldn't be able to update the actual transaction unless the vaccination was done on the same tablet...

Probably just need to agree on a 'source of truth' for this data. 

## Bugs found & Fixes

- [ ] #4785 
- [ ] #4791

## Implementation

**A description of how the solution would best be implemented in the code base**

## Describe alternatives you've considered

**A clear and concise description of any alternative solutions or features you've considered.**

## Additional context

**Add any other context or screenshots about the feature request here.**

### PR DESCRIPTION
**Pull request by katherine-sussol**
Fixes #4715

## Change summary

- Creates credit and new CI when someone tries to edit historical transaction data (allow users to change the vaccine type or the vaccinator - the batch is automatically selected for the user, thinking being no one is going to remember that far back...)
https://user-images.githubusercontent.com/65875762/176802817-f25ab475-14a1-4578-a9f2-634a7d0ba2e5.mp4
- Soft deletes old vaccination name note when edit is done
- Includes a slightly unhelpful popup when logged onto a different store than the one where the transaction was created (for now someone will have to manually run queries to figure out which tablet to use - going forward the store name will be saved into vaccination name notes)
https://user-images.githubusercontent.com/65875762/176801618-43382905-be0f-45c7-9739-c86b28799fdd.mp4

## Testing

Steps to reproduce or otherwise test the changes of this PR:

- [ ] Need a 'VaccinationEvent' form schema in Desktop - Used these ones
  - [VaccinationEvent.txt](https://github.com/openmsupply/mobile/files/9025098/VaccinationEvent.txt)
  - [VaccinationEvent-UI.txt](https://github.com/openmsupply/mobile/files/9025101/VaccinationEvent-UI.txt)
- [ ] Edit a vaccine that was dispensed at the same store
- [ ] Edit a vaccinator associated with a historical vaccination dispensed at the same store

### Related areas to think about

- The modal doesn't close or update when the transaction is complete (should probably be updated)
- Haven't tested whether name_note updates get propagated properly to other stores 
- There aren't currently any checks whether the item that is attempting to be returned to stock is still visible to the store (unsure what happens if someone removes visibility/stock)

### CODE REVIEW

**Review by mark-prins — APPROVED** [review]
👍 looks ok! am having trouble testing, and don't have the patience to wait for the kiribati sync which can take hours

File: src/actions/Entities/VaccinePrescriptionActions.js [hunk 1]
```diff
@@ -315,17 +319,33 @@ const confirmAndRepeat = () => dispatch =>
...
+  const { totalQuantity } = transactionBatch;
+  const patient = UIDatabase.get('Name', patientID);
+
+  UIDatabase.write(() => {
+    const customerCredit = createRecord(
+      UIDatabase,
+      'CustomerCredit',
+      currentUser,
+      patient,
+      -totalQuantity,
```
[the remark this answers is not in the document]
Comment (katherine-sussol) [reply src/actions/Entities/VaccinePrescriptionActions.js:334]: I wondered the same and originally updated the total/subtotal fields (referring to the supplierCredit implementation), thinking that 'total' meant total count of stock being moved, but that appears to actually be the $ amount in mSupply, so I removed it 😆 . I couldn't find an appropriate place to save it in the Transaction 🤔 , maybe better to just remove the misleading 'returnAmount' param?

File: src/database/utilities/createRecord.js [hunk 2]
```diff
@@ -532,39 +532,67 @@ const createOffsetCustomerCredit = (database, receipt) => {
...
   const refundLine = database.create('TransactionBatch', {
     id: generateUUID(),
-    item,
+    itemId: transactionItem.id,
+    itemName: transactionItem.name,
+    itemBatch,
     batch,
     expiryDate,
     packSize,
+    numberOfPacks: 0,
```
[the remark this answers is not in the document]
Comment (mark-prins) [reply src/database/utilities/createRecord.js:555]: actually.. nah. I've tested and the credit is created fine and it all works. ignore my comment

File: src/actions/Entities/VaccinePrescriptionActions.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (mark-prins) [reply src/actions/Entities/VaccinePrescriptionActions.js:334]: actually, it's all good. objection withdrawn.

File: src/widgets/modals/VaccinationEvent.js [hunk 3]
```diff
@@ -46,18 +51,36 @@ export const NoPCDForm = () => (
...
+  vaccinationEventId,
+  vaccinationEventSchema,
+  vaccines,
 }) => {
   const pcdFormRef = useRef(null);
   const supplementalFormRef = useRef(null);
   const vaccinationFormRef = useRef(null);
 
+  const vaccinationEventNameNote = UIDatabase.get('NameNote', vaccinationEventId);
+  const vaccinationEvent = vaccinationEventNameNote.data;
```
Comment (sworup) [inline src/widgets/modals/VaccinationEvent.js:69]: Getting this error while testing in emulator.

Screen Shot 2022-07-01 at 11 59 14 AM

Comment (sworup) [reply src/widgets/modals/VaccinationEvent.js:69]: Some other places this shows up too. Basically we need to validate notes for cases where there is no fields.

**Review by sworup — COMMENTED** [review]
In case the patient does not have proper data in NameNote the app crashes.

File: src/widgets/modals/VaccinationEvent.js (hunk 3 above)
Comment (sworup) [reply src/widgets/modals/VaccinationEvent.js:69]: Screen Shot 2022-07-01 at 12 27 49 PM

**Comment by sworup** [comment]
Found some other bugs.


https://user-images.githubusercontent.com/683452/176842536-0cfe12c0-420a-4379-8ab3-085fbf9b01b9.mp4

**Comment by anildahalsussol** [comment]
For me, while saving the changes in the extra information is not updated

**Comment by chastichno** [comment]
@sworup @anildahalsussol I just tested and see the same - changes are not saved sometimes, behaviour of "Save changes" buttons is not consistent. I suggest we merge current changes and create new issue with information about all the found bugs

**Comment by sworup** [comment]
> @sworup @anildahalsussol I just tested and see the same - changes are not saved sometimes, behaviour of "Save changes" buttons is not consistent. I suggest we merge current changes and create new issue with information about all the found bugs

Sounds good @chastichno.

**Comment by chastichno** [comment]
I just tested it again and everything seems to be working for me. The only problem left is the confusing UI - after I edit the vaccinator or vaccine there's no confirmation and the "editTransaction" block doesn't close.
My suggestion is:
- to show a confirmation window after clicking "Save changes".
- to show an alert after changes are successfully saved saying that the vaccination transaction was edited (might be redundant, but changing a finalised invoice IS a big deal sooo why not make it look very serious). I also suggest we disable the "Edit" button after that as it can cause bugs in the database - if you click "Edit" again and edit info again it will create a customer credit for the old vaccination. If there's a need user can always change a newly created vaccination.
@mark-prins What do you think?
The current PR can be merged though 👍

**Review by chastichno — APPROVED** [review]

**Pull request merged**